google/cloud/bigtable/admin/internal/wait_for_consistency.cc

```diff
@@ -0,0 +1,184 @@
+// limitations under the License.
+
+#include "google/cloud/bigtable/admin/internal/wait_for_consistency.h"
+#include "google/cloud/bigtable/admin/bigtable_table_admin_client.h"
+#include "google/cloud/bigtable/admin/bigtable_table_admin_options.h"
+#include "google/cloud/bigtable/admin/internal/bigtable_table_admin_option_defaults.h"
+#include <chrono>
+
+namespace google {
+namespace cloud {
+namespace bigtable_admin_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+using RespType = StatusOr<bigtable::admin::v2::CheckConsistencyResponse>;
+using TimerFuture = future<StatusOr<std::chrono::system_clock::time_point>>;
+
+// This class borrows heavily from `google::cloud::internal::AsyncRetryLoop`
+class AsyncWaitForConsistencyImpl
+    : public std::enable_shared_from_this<AsyncWaitForConsistencyImpl> {
+ public:
+  AsyncWaitForConsistencyImpl(CompletionQueue cq,
+                              bigtable_admin::BigtableTableAdminClient client,
+                              std::string table_name,
+                              std::string consistency_token,
+                              Options const& options)
+      : cq_(std::move(cq)),
+        client_(std::move(client)),
+        table_name_(std::move(table_name)),
+        consistency_token_(std::move(consistency_token)),
+        options_(BigtableTableAdminDefaultOptions(std::move(options))),
+        polling_policy_(
+            options_
+                .get<bigtable_admin::BigtableTableAdminPollingPolicyOption>()) {
+  }
+
+  future<Status> Start() {
+    auto w = std::weak_ptr<AsyncWaitForConsistencyImpl>(shared_from_this());
+    result_ = promise<Status>([w] {
+      if (auto self = w.lock()) self->Cancel();
+    });
+
+    StartAttempt();
+    return result_.get_future();
+  }
+
+ private:
+  enum State {
+    kIdle,
+    kWaiting,
+    kDone,
+  };
+
+  void StartAttempt() {
+    auto self = this->shared_from_this();
+    auto op =
+        client_.AsyncCheckConsistency(table_name_, consistency_token_, options_)
+            .then([self](future<RespType> f) { self->OnAttempt(f.get()); });
+    SetWaiting(std::move(op));
+  }
+
+  void OnAttempt(RespType result) {
+    SetIdle();
+    // A successful attempt, set the value and finish the loop.
+    if (result.ok() && result->consistent()) {
+      SetDone(Status());
+      return;
+    }
+    auto status = std::move(result).status();
+    if (!polling_policy_->OnFailure(status)) {
+      if (status.ok()) {
+        SetDone(Status(StatusCode::kDeadlineExceeded,
+                       "Polling loop terminated by polling policy"));
+      } else {
+        SetDone(std::move(status));
+      }
+      return;
+    }
+    if (Cancelled()) return;
+    auto self = this->shared_from_this();
+    auto op =
+        cq_.MakeRelativeTimer(polling_policy_->WaitPeriod())
+            .then([self](TimerFuture f) { self->OnBackoffTimer(f.get()); });
+    SetWaiting(std::move(op));
```

Comment:
As above.  The OS scheduler could insert a long sleep between lines 95 and 96, and by the time you call `SetWaiting()` then the timer has expired, and a new operation started.  There is an inherent race condition, it seems you need a mutex around this critical section.